### ISSUE TITLE
http2: add per connection(stream 0) HTTP/2 METADATA frames

### ISSUE THREAD

**Issue by lambdai** [envoyproxy/envoy#15608]
*Description*:
Envoy added per stream metadata frame #2394

It's a naive requirement to extend the METADATA frame support to stream 0.

Below metadata fit per connection metadata category. 

1. NodeInfo: the node info is supposed to be stable during the life of the connection. 
2. Connection trace context. 

Where to put the new config?
A `ConnectionMetaData` message and field within `Http2ProtocolOptions`. This option lives in both client and server http/2 connections.

Is anyone feel this useful?

CC @htuch @alyssawilk

**Comment by lambdai** [comment]
CC @tbarrella This might affect the **eventual** shape of our upstream cluster config

**Comment by alyssawilk** [comment]
I don't think any of the teams currently using metadata have a need for it.
Can you flesh out the use case a bit more? Who would receive that metadata and what would they do with it

**Comment by lambdai** [comment]
> I don't think any of the teams currently using metadata have a need for it.
> Can you flesh out the use case a bit more? Who would receive that metadata and what would they do with it

stream 0 metadata is going to address issue that per stream metadata is sent again and again.

Assume the both end is Envoy,
1. The async client can encode the client node info into the stream 0. The server side HCM can create fake stream 0 to decode metadata and put in conneciton stream info, so that stream 1,3,... can get these per connection info without decoding. 
2. For the metadata from server node should have the similar op. Although there is no http upstream filter yet, we can still persist the stream 0 metadata from server in the http async client. Ideally each upstream request could consume the metadata.

Alternatively we can specialize ConnectionImpl::onMetadataReceived if the passed in stream id == 0.

The metadata by definition is not a req/resp mode. It's not too hard to define a per connection metadata filter spec and only allow the metadata filter to access stream info and H2ProtocolOptions.

Does the above motivation reasonable?

**Comment by alyssawilk** [comment]
"stream 0 metadata is going to address issue that per stream metadata is sent again and again."

Again, what stream metadata is sent again and again?  I don't think I understand the use case yet.

The use cases I know of metadata are to send things like RTT, which changes over the course of the stream and session lifetime so it's useful to resend.  Alternately if you want to send once and have it resent on all further streams, you could do something like have an HTTP filter which intercepted the metadata, added it to the connection's StreamInfo, and other streams would pull it from the connection's StreamInfo and forward it on as metadata.  I don't think we'd need HCM changes for that.  WDYT?

**Comment by lambdai** [comment]
> Alternately if you want to send once and have it resent on all further streams, you could do something like have an HTTP filter which intercepted the metadata, added it to the connection's StreamInfo, and other streams would pull it from the connection's StreamInfo and forward it on as metadata.

Thanks!  This kind of metadata is shared by all streams within the parent connection, ideally we want a dedicate stream 0 which follows how stream 0 is defined.
I think this should work: The first stream set and the rest of the rest of the streams consume to avoid HCM change

**Comment by mattklein123** [comment]
@lambdai per Alyssa, do you have an actual use case for this? If so, what is it? We don't want to add a feature like this unless there is a real use case that someone is going to implement.

**Comment by lambdai** [comment]
The use case is to transmit connection attributes and node attributes bidirectionally.  

I see node metadata and connection metadata can be delivered without metadata but using header. 

stream 0 is easier to build "wait for certain metadata" before the normal http stream, but the first stream to carry and set should work. Please consider the stream 0 metadata as "nice to have" and we can wait for more use cases

**Comment by mattklein123** [comment]
I understand what it can theoretically do, but what I want to know is if someone has a production use case they are going to implement. Until that exists let's hold on this.

**Comment by lambdai** [comment]
> I understand what it can theoretically do, but what I want to know is if someone has a production use case they are going to implement. Until that exists let's hold on this.

Istio does use https://github.com/istio/proxy/tree/master/src/envoy/tcp/metadata_exchange as a workaround. Think of it as an extended PROXY protocol.  
Istio also has http version which currently repeatedly exchanges metadata for each stream. 

Also istio is pushing all tcp traffic to be tunnelled in http/2, so per connection metadata serve both http traffic and tcp traffic in istio world.

I create this issue to see if other people find per connection metadata a better abstraction to replace the individual workarounds in production

**Comment by alyssawilk** [comment]
Is the per connection data something you want to send once at the beginning of the connection, or something you need optional updates on?  You said node info remains constant and I'm not sure if you'd be passing actual trace events, or just a trace ID.

While conceptually I like the idea of connection metadata on stream 0, one problem offhand with doing this at the HTTP/2 layer, is where we'd stick plug-ins. If we do something like proxy proto we could ideally have tcp upstream filters adding this metadata, and tcp downstream filters interpreting it and putting it in stream info, and folks can customize what they'd like to have added and removed.  But if the HCM is getting per connection data, it's unclear where we'd plug in the per-customer business logic of what data is added and how it is processed. Plus if we allow "proxy proto extensibility" it's useful for both HTTP/2-over-tcp and actual TCP (google has our own in-house proxy proto style header).

**Comment by mattklein123** [comment]
Yeah it does seem like we would need stream 0/connection filters for this kind of thing. Overall I like this idea, but we should flesh it out into a full design doc I think.

**Comment by lambdai** [comment]
> Is the per connection data something you want to send once at the beginning of the connection, or something you need optional updates on? You said node info remains constant and I'm not sure if you'd be passing actual trace events, or just a trace ID.

The trace id is what I was referring to. All of my existing production use cases are send-once-shared-by-all-stream.
Delivering occasional trace event during the full life cycle of h2 connection is attractive but I don't have a timeline for this requirement. The stream 0 filter api is obscure for this scenario. I believe a congestion control is a hard dependency.

I prefer build the "metadata exchange in h2" to "L4 transport" because h2 provides the concept of metadata frame and per connection stream(stream 0) and per business stream (1,3,...).

The PROXY protocol was a pain because I see PROXY protocol as degraded version of "metadata stream": It only allows transmit at beginning, it's require the receiver side aware of the header.
If we extend the PROXY protocol-ish thing we may eventually end up with a subset of h2.

**Comment by alyssawilk** [comment]
Ok, so right now and for the foreseeable future you only need to transmit data at the beginning of the connection.  

for the receive-side-aware argument, if you try to negotiate h2-with-medata and the peer rejects it, is Enovy savvy enough to not send metadata, or does it fail to connect entirely?  I would think it would fail to connect, or send invalid frames at which point both solutions have the same issue where the receive side needs data.

Either way if you really feel stream 0 metadata is the best solution can you do a one pager with suggested APIs?  I still think there's going to be a problem having a general extensible solution to L4 metadata being added and removed from L7 streams cleanly which I think we should sort out before going forward.

**Comment by lambdai** [comment]
> Ok, so right now and for the foreseeable future you only need to transmit data at the beginning of the connection.
> 
Thank you for the help to clarify it!

> for the receive-side-aware argument, if you try to negotiate h2-with-medata and the peer rejects it, is Enovy savvy enough to not send metadata, or does it fail to connect entirely? I would think it would fail to connect, or send invalid frames at which point both solutions have the same issue where the receive side needs data.
> 

I wish the control plane can deliver if the metadata is expected at the peer to avoid handshake. The handshake adds roundtrip and complexity of state machine. Meanwhile,  either Envoy as client or Envoy as server should not crash (need to verify but I trust our existing code) when unexpected metadata receives. 
I have two cases: 
In one case the rest of the streams(1,3...) require the metadata, so the filter which consumes received metadata should reset the connection.
In the other case, the rest of the streams(1,3...)  see the metadata as optional, e.g. telemetry metadata. 
It will be reflected in the design doc when the doc is ready.


> Either way if you really feel stream 0 metadata is the best solution can you do a one pager with suggested APIs? I still think there's going to be a problem having a general extensible solution to L4 metadata being added and removed from L7 streams cleanly which I think we should sort out before going forward.

Yeah... I agree we can either lay the metadata down to L4 protocol. It is proved to be somewhat workable in most of the cases but failed on certain weird and picky L4 firewalls.
I will ping you when the doc is ready. 

Thank you all!
